// thirdweb/dist/cjs/extensions/ens/resolve-name.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveName = resolveName;
const ethereum_js_1 = require("../../chains/chain-definitions/ethereum.js");
const contract_js_1 = require("../../contract/contract.js");
const withCache_js_1 = require("../../utils/promise/withCache.js");
const reverse_js_1 = require("./__generated__/UniversalResolver/read/reverse.js");
const constants_js_1 = require("./constants.js");
/**
 * Resolves the primary name for a specified address.
 * @param options - The options for resolving an ENS address.
 * @example
 * ```ts
 * import { resolveName } from "thirdweb/extensions/ens";
 * const name = await resolveName({
 *    client,
 *    address: "0x1234...",
 * });
 * ```
 * @extension ENS
 * @returns A promise that resolves to the Ethereum address.
 */
async function resolveName(options) {
    const { client, address, resolverAddress, resolverChain } = options;
    return (0, withCache_js_1.withCache)(async () => {
        const contract = (0, contract_js_1.getContract)({
            address: resolverAddress || constants_js_1.UNIVERSAL_RESOLVER_ADDRESS,
            chain: resolverChain || ethereum_js_1.ethereum,
            client,
        });
        const [name] = await (0, reverse_js_1.reverse)({
            contract,
            reverseName: address,
            coinType: 60n,
        }).catch((e) => {
            // Re-throw verification errors so callers can detect data integrity issues
            if (typeof e === "object" &&
                e !== null &&
                "data" in e &&
                typeof e.data === "string") {
                // ReverseAddressMismatch(string,bytes) = 0xef9c03ce
                if (e.data.startsWith("0xef9c03ce")) {
                    throw e;
                }
            }
            // Swallow expected "no resolver" / "no name" errors
            return [null];
        });
        return name || null;
    }, {
        cacheKey: `ens:name:${resolverChain?.id || 1}:${address}`,
        // 1min cache
        cacheTime: 60 * 1000,
    });
}
